File: backend/pkg/chunking/json/json.go
package json

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"github.com/OFFIS-RIT/kiwi/backend/pkg/chunking"
	gonanoid "github.com/matoous/go-nanoid/v2"
	"github.com/pkoukk/tiktoken-go"
)

// JSONChunker implements a 3-tier chunking strategy for JSON values:
//
//  1. If the entire JSON fits within maxChunkSize tokens, return it as a single chunk.
//  2. Otherwise, split by top-level keys (objects) or top-level elements (arrays),
//     grouping entries that fit together into one chunk.
//  3. If any single top-level entry still exceeds the limit, recursively split
//     its children with path-aware context prefixes.
type JSONChunker struct {
	maxChunkSize int
	encoder      string
}

type NewJSONChunkerParams struct {
	MaxChunkSize int
	Encoder      string
}

func NewJSONChunker(params NewJSONChunkerParams) *JSONChunker {
	return &JSONChunker{
		maxChunkSize: params.MaxChunkSize,
		encoder:      params.Encoder,
	}
}

func (c *JSONChunker) Chunk(_ context.Context, input string) ([]chunking.Chunk, error) {
	enc, err := tiktoken.GetEncoding(c.encoder)
	if err != nil {
		return nil, err
	}

	text := strings.TrimSpace(input)
	if text == "" {
		return nil, nil
	}

	tokenCount := func(s string) int {
		return len(enc.Encode(s, nil, nil))
	}

	// Tier 1: entire document fits in one chunk.
	if tokenCount(text) <= c.maxChunkSize {
		id, err := gonanoid.New()
		if err != nil {
			return nil, err
		}
		return []chunking.Chunk{{ID: id, Text: text}}, nil
	}

	// Parse the JSON to determine structure.
	var raw any
	if err := json.Unmarshal([]byte(text), &raw); err != nil {
		// Not valid JSON — fall back to returning as a single chunk.
		id, err := gonanoid.New()
		if err != nil {
			return nil, err
		}
		return []chunking.Chunk{{ID: id, Text: text}}, nil
	}

	var textChunks []string

	switch v := raw.(type) {
	case map[string]any:
		textChunks, err = c.chunkObject(v, "$", orderedKeys([]byte(text)), tokenCount)
	case []any:
		textChunks, err = c.chunkArray(v, "$", tokenCount)
	default:
		// Scalar at top level — single chunk.
		id, err := gonanoid.New()
		if err != nil {
			return nil, err
		}
		return []chunking.Chunk{{ID: id, Text: text}}, nil
	}

	if err != nil {
		return nil, err
	}

	result := make([]chunking.Chunk, len(textChunks))
	for i, t := range textChunks {
		id, err := gonanoid.New()
		if err != nil {
			return nil, err
		}
		result[i] = chunking.Chunk{ID: id, Text: t}
	}

	return result, nil
}

// chunkObject splits a JSON object by its keys, grouping keys that fit together
// into one chunk. For top-level objects, keyOrder should come from the original
// JSON text. If empty, deterministic lexical order is used.
func (c *JSONChunker) chunkObject(
	obj map[string]any,
	path string,
	keyOrder []string,
	tokenCount func(string) int,
) ([]string, error) {
	keys := objectKeysInOrder(obj, keyOrder)

	var chunks []string
	currentEntries := make(map[string]any)
	currentTokens := 0

	flush := func() error {
		if len(currentEntries) == 0 {
			return nil
		}
		text, err := prettyMarshal(currentEntries)
		if err != nil {
			return err
		}
		if path != "$" {
			text = fmt.Sprintf("Path: %s\n%s", path, text)
		}
		chunks = append(chunks, text)
		currentEntries = make(map[string]any)
		currentTokens = 0
		return nil
	}

	for _, key := range keys {
		val, ok := obj[key]
		if !ok {
			continue
		}
		entry := map[string]any{key: val}
		entryText, err := prettyMarshal(entry)
		if err != nil {
			return nil, err
		}
		entryTokens := tokenCount(entryText)

		// Tier 3: single entry exceeds the limit — recurse into its value.
		if entryTokens > c.maxChunkSize {
			// Flush any accumulated entries first.
			if err := flush(); err != nil {
				return nil, err
			}
			childPath := fmt.Sprintf("%s.%s", path, key)
			subChunks, err := c.chunkValue(val, childPath, tokenCount)
			if err != nil {
				return nil, err
			}
			chunks = append(chunks, subChunks...)
			continue
		}

		// Tier 2: accumulate entries that fit together.
		if currentTokens+entryTokens > c.maxChunkSize && len(currentEntries) > 0 {
			if err := flush(); err != nil {
				return nil, err
			}
		}

		currentEntries[key] = val
		currentTokens += entryTokens
	}

	if err := flush(); err != nil {
		return nil, err
	}

	return chunks, nil
}

// chunkArray splits a JSON array by batching elements that fit together.
// If a single element exceeds the limit, it is recursively split.
func (c *JSONChunker) chunkArray(
	arr []any,
	path string,
	tokenCount func(string) int,
) ([]string, error) {
	var chunks []string
	var currentElements []any
	currentTokens := 0

	flush := func() error {
		if len(currentElements) == 0 {
			return nil
		}
		text, err := prettyMarshal(currentElements)
		if err != nil {
			return err
		}
		if path != "$" {
			text = fmt.Sprintf("Path: %s\n%s", path, text)
		}
		chunks = append(chunks, text)
		currentElements = nil
		currentTokens = 0
		return nil
	}

	for i, elem := range arr {
		elemText, err := prettyMarshal(elem)
		if err != nil {
			return nil, err
		}
		elemTokens := tokenCount(elemText)

		// Single element exceeds limit — recurse.
		if elemTokens > c.maxChunkSize {
			if err := flush(); err != nil {
				return nil, err
			}
			childPath := fmt.Sprintf("%s[%d]", path, i)
			subChunks, err := c.chunkValue(elem, childPath, tokenCount)
			if err != nil {
				return nil, err
			}
			chunks = append(chunks, subChunks...)
			continue
		}

		if currentTokens+elemTokens > c.maxChunkSize && len(currentElements) > 0 {
			if err := flush(); err != nil {
				return nil, err
			}
		}

		currentElements = append(currentElements, elem)
		currentTokens += elemTokens
	}

	if err := flush(); err != nil {
		return nil, err
	}

	return chunks, nil
}

// chunkValue dispatches to the appropriate chunker based on value type.
// Scalar values or values that fit within the limit are returned as-is with a path prefix.
func (c *JSONChunker) chunkValue(
	val any,
	path string,
	tokenCount func(string) int,
) ([]string, error) {
	switch v := val.(type) {
	case map[string]any:
		return c.chunkObject(v, path, nil, tokenCount)
	case []any:
		return c.chunkArray(v, path, tokenCount)
	default:
		// Scalar — return as a single chunk with path context.
		text, err := prettyMarshal(val)
		if err != nil {
			return nil, err
		}
		return []string{fmt.Sprintf("Path: %s\n%s", path, text)}, nil
	}
}

func objectKeysInOrder(obj map[string]any, preferred []string) []string {
	if len(preferred) == 0 {
		keys := make([]string, 0, len(obj))
		for key := range obj {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		return keys
	}

	keys := make([]string, 0, len(obj))
	seen := make(map[string]struct{}, len(obj))
	for _, key := range preferred {
		if _, ok := obj[key]; !ok {
			continue
		}
		if _, ok := seen[key]; ok {
			continue
		}
		keys = append(keys, key)
		seen[key] = struct{}{}
	}

	if len(keys) == len(obj) {
		return keys
	}

	missing := make([]string, 0, len(obj)-len(keys))
	for key := range obj {
		if _, ok := seen[key]; ok {
			continue
		}
		missing = append(missing, key)
	}
	sort.Strings(missing)

	return append(keys, missing...)
}

// orderedKeys extracts key order from JSON object bytes.
func orderedKeys(data []byte) []string {
	dec := json.NewDecoder(strings.NewReader(string(data)))

	// Read opening '{'
	t, err := dec.Token()
	if err != nil {
		return nil
	}
	if delim, ok := t.(json.Delim); !ok || delim != '{' {
		return nil
	}

	var keys []string
	for dec.More() {
		t, err := dec.Token()
		if err != nil {
			break
		}
		if key, ok := t.(string); ok {
			keys = append(keys, key)
			// Skip the value.
			var skip json.RawMessage
			if err := dec.Decode(&skip); err != nil {
				break
			}
		}
	}

	return keys
}

// prettyMarshal serializes a value as indented JSON.
func prettyMarshal(v any) (string, error) {
	b, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return "", err
	}
	return string(b), nil
}

File: packages/graph/src/chunking/json.ts
import type { GraphChunker } from "..";
import { get_encoding } from "tiktoken";

type JSONChunkerOptions = {
    maxChunkSize: number;
    encoder?: string;
};

type JSONValue = null | boolean | number | string | JSONValue[] | { [key: string]: JSONValue };

export class JSONChunker implements GraphChunker {
    private readonly maxChunkSize: number;
    private readonly encoderName: string;

    constructor(options: JSONChunkerOptions) {
        this.maxChunkSize = options.maxChunkSize;
        this.encoderName = options.encoder ?? "o200k_base";
    }

    async getChunks(input: string): Promise<string[]> {
        const text = input.trim();
        if (text === "") {
            return [];
        }

        const encoder = get_encoding(this.encoderName as Parameters<typeof get_encoding>[0]);

        try {
            const tokenCount = (value: string) => encoder.encode(value).length;

            if (tokenCount(text) <= this.maxChunkSize) {
                return [text];
            }

            let raw: JSONValue;
            try {
                raw = JSON.parse(text) as JSONValue;
            } catch {
                return [text];
            }

            if (isJSONObject(raw)) {
                return this.chunkObject(raw, "$", orderedKeys(text), tokenCount);
            }

            if (Array.isArray(raw)) {
                return this.chunkArray(raw, "$", tokenCount);
            }

            return [text];
        } finally {
            encoder.free();
        }
    }

    private chunkObject(
        obj: Record<string, JSONValue>,
        path: string,
        keyOrder: string[] | undefined,
        tokenCount: (text: string) => number
    ): string[] {
        const keys = objectKeysInOrder(obj, keyOrder);
        const chunks: string[] = [];
        let currentEntries: Record<string, JSONValue> = {};
        let currentTokens = 0;

        const flush = () => {
            if (Object.keys(currentEntries).length === 0) {
                return;
            }

            let text = prettyStringify(currentEntries);
            if (path !== "$") {
                text = `Path: ${path}\n${text}`;
            }

            chunks.push(text);
            currentEntries = {};
            currentTokens = 0;
        };

        for (const key of keys) {
            if (!(key in obj)) {
                continue;
            }

            const value = obj[key]!;

            const entry: Record<string, JSONValue> = { [key]: value };
            const entryText = prettyStringify(entry);
            const entryTokens = tokenCount(entryText);

            if (entryTokens > this.maxChunkSize) {
                flush();
                chunks.push(...this.chunkValue(value, `${path}.${key}`, tokenCount));
                continue;
            }

            if (currentTokens + entryTokens > this.maxChunkSize && Object.keys(currentEntries).length > 0) {
                flush();
            }

            currentEntries[key] = value;
            currentTokens += entryTokens;
        }

        flush();
        return chunks;
    }

    private chunkArray(values: JSONValue[], path: string, tokenCount: (text: string) => number): string[] {
        const chunks: string[] = [];
        let currentValues: JSONValue[] = [];
        let currentTokens = 0;

        const flush = () => {
            if (currentValues.length === 0) {
                return;
            }

            let text = prettyStringify(currentValues);
            if (path !== "$") {
                text = `Path: ${path}\n${text}`;
            }

            chunks.push(text);
            currentValues = [];
            currentTokens = 0;
        };

        values.forEach((value, index) => {
            const entryText = prettyStringify(value);
            const entryTokens = tokenCount(entryText);

            if (entryTokens > this.maxChunkSize) {
                flush();
                chunks.push(...this.chunkValue(value, `${path}[${index}]`, tokenCount));
                return;
            }

            if (currentTokens + entryTokens > this.maxChunkSize && currentValues.length > 0) {
                flush();
            }

            currentValues.push(value);
            currentTokens += entryTokens;
        });

        flush();
        return chunks;
    }

    private chunkValue(value: JSONValue, path: string, tokenCount: (text: string) => number): string[] {
        if (isJSONObject(value)) {
            return this.chunkObject(value, path, undefined, tokenCount);
        }

        if (Array.isArray(value)) {
            return this.chunkArray(value, path, tokenCount);
        }

        return [`Path: ${path}\n${prettyStringify(value)}`];
    }
}

function isJSONObject(value: JSONValue): value is Record<string, JSONValue> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function objectKeysInOrder(obj: Record<string, JSONValue>, preferred: string[] | undefined): string[] {
    if (!preferred || preferred.length === 0) {
        return Object.keys(obj).sort();
    }

    const seen = new Set<string>();
    const ordered: string[] = [];

    for (const key of preferred) {
        if (!(key in obj) || seen.has(key)) {
            continue;
        }

        ordered.push(key);
        seen.add(key);
    }

    if (ordered.length === Object.keys(obj).length) {
        return ordered;
    }

    const missing = Object.keys(obj)
        .filter((key) => !seen.has(key))
        .sort();

    return [...ordered, ...missing];
}

function orderedKeys(input: string): string[] | undefined {
    const trimmed = input.trim();
    if (!trimmed.startsWith("{")) {
        return undefined;
    }

    try {
        const keys: string[] = [];
        let index = 0;

        const skipWhitespace = () => {
            while (index < trimmed.length && /\s/u.test(trimmed[index]!)) {
                index += 1;
            }
        };

        const readString = (): string | undefined => {
            if (trimmed[index] !== '"') {
                return undefined;
            }

            index += 1;
            let value = "";
            let escaped = false;

            while (index < trimmed.length) {
                const char = trimmed[index];
                index += 1;

                if (escaped) {
                    value += `\\${char}`;
                    escaped = false;
                    continue;
                }

                if (char === "\\") {
                    escaped = true;
                    continue;
                }

                if (char === '"') {
                    return JSON.parse(`"${value}"`) as string;
                }

                value += char;
            }

            return undefined;
        };

        const skipValue = (): boolean => {
            skipWhitespace();
            const start = index;
            let depth = 0;
            let inString = false;
            let escaped = false;

            while (index < trimmed.length) {
                const char = trimmed[index];

                if (inString) {
                    index += 1;
                    if (escaped) {
                        escaped = false;
                    } else if (char === "\\") {
                        escaped = true;
                    } else if (char === '"') {
                        inString = false;
                    }
                    continue;
                }

                if (char === '"') {
                    inString = true;
                    index += 1;
                    continue;
                }

                if (char === "{" || char === "[") {
                    depth += 1;
                    index += 1;
                    continue;
                }

                if (char === "}" || char === "]") {
                    if (depth === 0) {
                        return index > start;
                    }
                    depth -= 1;
                    index += 1;
                    continue;
                }

                if (depth === 0 && char === ",") {
                    return index > start;
                }

                index += 1;
            }

            return index > start;
        };

        skipWhitespace();
        if (trimmed[index] !== "{") {
            return undefined;
        }
        index += 1;

        while (index < trimmed.length) {
            skipWhitespace();
            if (trimmed[index] === "}") {
                break;
            }

            const key = readString();
            if (key === undefined) {
                return undefined;
            }
            keys.push(key);

            skipWhitespace();
            if (trimmed[index] !== ":") {
                return undefined;
            }
            index += 1;

            if (!skipValue()) {
                return undefined;
            }

            skipWhitespace();
            if (trimmed[index] === ",") {
                index += 1;
            }
        }

        return keys;
    } catch {
        return undefined;
    }
}

function prettyStringify(value: JSONValue): string {
    return JSON.stringify(value, null, 2);
}
